core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/util/QueryEvaluationUtil.java

```diff
@@ -123,9 +129,48 @@ private static boolean valuesEqual(Value leftVal, Value rightVal) {
 		return leftVal != null && rightVal != null && leftVal.equals(rightVal);
 	}
 
+	/**
+	 * Compares the supplied {@link Literal} arguments using the supplied operator, using strict
+	 * (minimally-conforming) SPARQL 1.1 operator behavior.
+	 * 
+	 * @param leftLit
+	 *        the left literal argument of the comparison.
+	 * @param rightLit
+	 *        the right literal argument of the comparison.
+	 * @param operator
+	 *        the comparison operator to use.
+	 * @return {@code true} if execution of the supplied operator on the supplied arguments succeeds,
+	 *         {@code false} otherwise.
+	 * @throws ValueExprEvaluationException
+	 *         if a type error occurred.
+	 */
 	public static boolean compareLiterals(Literal leftLit, Literal rightLit, CompareOp operator)
 		throws ValueExprEvaluationException
 	{
+		return compareLiterals(leftLit, rightLit, operator, true);
+	}
+
+	/**
+	 * Compares the supplied {@link Literal} arguments using the supplied operator.
+	 * 
+	 * @param leftLit
+	 *        the left literal argument of the comparison.
+	 * @param rightLit
+	 *        the right literal argument of the comparison.
+	 * @param operator
+	 *        the comparison operator to use.
+	 * @param strict
+	 *        boolean indicating whether comparison should use strict (minimally-conforming) SPARQL 1.1
+	 *        operator behavior, or extended behavior.
+	 * @return {@code true} if execution of the supplied operator on the supplied arguments succeeds,
+	 *         {@code false} otherwise.
+	 * @throws ValueExprEvaluationException
+	 *         if a type error occurred.
+	 */
+	public static boolean compareLiterals(Literal leftLit, Literal rightLit, CompareOp operator,
+			boolean strict)
```

Comment:
I generally dislike the use of a boolean flag in this context too, but since there is a lot of common code between the two flavours, it seems reasonable to me. To avoid it, you would probably have to turn that method into a class with fine-grained methods that could be overridden to extend the comparison behaviour. Or perhaps have the method return something like an Optional<Boolean>, notPresent being return when there is no commonDataType. A caller could then check for this to decide whether to delegate the comparison to some compareExtendedLiterals() method, e.g. caller code could look something like: return coalesce(compareStrictLiterals(...), compareExtendedLiterals(...)